<!DOCTYPE html>
<html lang="en">
	<head>
		
		
		<meta charset="utf-8">
		<meta http-equiv="X-UA-Compatible" content="IE=edge">
		<meta name="viewport" content="width=device-width, initial-scale=1.0">
		
		<meta name="author" content="Jacques Dainat et al.">
		<link rel="canonical" href="https://Juke34.github.io/workshop-reproducible-research/snakemake/snakemake-5-parameters/">
		<link rel="shortcut icon" href="../../img/favicon.ico">
		<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" />
		<title>Parameters - Bioinformatics Workshop on Tools for Reproducible Research</title>
		<link href="../../bootstrap/dist/css/bootstrap.min.css" rel="stylesheet">
		<link href="../../css/font-awesome-5.12.2-all.min.css" rel="stylesheet">
		<link href="../../css/base.css" rel="stylesheet">
		<link rel="stylesheet" href="../../css/highlight-github.css">
		<script src="../../bootstrap/dist/js/bootstrap.min.js"></script>
		<script src="../../js/highlight.pack.js"></script>
		
		<script>
			var base_url = '../..';
			var home_url = '../../usage';
			var is_outer_page = false;
			
			var pageToc = [
			];

		</script>
		<script src="../../js/base.js"></script> 
	</head>

	<body class="inner-page">
		

		<div class="container-fluid wm-page-content">
			<a name="_top" aria-hidden="true"></a>
			






<div class="row pt-2 pb-4">
	<div class="col-sm text-center text-sm-left">
		<a href="../snakemake-4-the-mrsa-workflow/" class="btn btn-sm btn-outline-dark">
			<i class="fa fa-chevron-left" aria-hidden="true"></i>
			Previous
		</a>
		<a href="../snakemake-4-the-mrsa-workflow/" class="btn btn-sm btn-link">
			The MRSA Workflow
		</a>
	</div>
	

	
	<div class="col-sm text-center text-sm-right">
		<a href="../snakemake-6-logs/" class="btn btn-sm btn-link">
			Logs
		</a>
		<a href="../snakemake-6-logs/" class="btn btn-sm btn-outline-dark">
			Next
			<i class="fa fa-chevron-right" aria-hidden="true"></i>
		</a>
	</div>
	
</div>

			

			<p>In a typical bioinformatics project, considerable efforts are spent on tweaking
parameters for the various programs involved. It would be inconvenient if you
had to change in the shell scripts themselves every time you wanted to run with
a new setting. Luckily, there is a better option for this: the <code>params</code>
keyword.</p>
<pre><code class="language-python">rule some_rule:
    output:
        &quot;...&quot;
    input:
        &quot;...&quot;
    params:
        cutoff=2.5
    shell:
        &quot;&quot;&quot;
        some_program --cutoff {params.cutoff} {input} {output}
        &quot;&quot;&quot;
</code></pre>
<p>We run most of the programs with default settings in our workflow. However,
there is one parameter in the rule <code>get_SRA_by_accession</code> that we use for
determining how many reads we want to retrieve from SRA for each sample 
(<code>-X 25000</code>). Change in this rule to use the parameter <code>max_reads</code> instead and 
set the value to 20000. If you need help, click to show the solution below.</p>
<details>
<summary> Click to show </summary>



<pre><code class="language-python">rule get_SRA_by_accession:
    &quot;&quot;&quot;
    Retrieve a single-read FASTQ file from SRA (Sequence Read Archive) by run accession number.
    &quot;&quot;&quot;
    output:
        &quot;data/raw_internal/{sample_id}.fastq.gz&quot;
    params:
        max_reads = 20000
    shell:
        &quot;&quot;&quot;
        fastq-dump {wildcards.sample_id} -X {params.max_reads} --readids \
            --dumpbase --skip-technical --gzip -Z &gt; {output}
        &quot;&quot;&quot;
</code></pre>


</details>

<p>Now run through the workflow. Because there's been changes to the <code>get_SRA_by_accession</code>
rule this will trigger a re-run of the rule for all three accessions. In addition
all downstream rules that depend on output from <code>get_SRA_by_accession</code> are re-run. </p>
<p>The parameter values we set in the <code>params</code> section don't have to be static,
they can be any Python expression. In particular, Snakemake provides a global
dictionary of configuration parameters called <code>config</code>. Let's modify
<code>get_SRA_by_accession</code> to look something like this in order to make use of this 
dictionary:</p>
<pre><code class="language-python">rule get_SRA_by_accession:
    &quot;&quot;&quot;
    Retrieve a single-read FASTQ file from SRA (Sequence Read Archive) by run accession number.
    &quot;&quot;&quot;
    output:
        &quot;data/raw_internal/{sample_id}.fastq.gz&quot;
    params:
        max_reads = config[&quot;max_reads&quot;]
    shell:
        &quot;&quot;&quot;
        fastq-dump {wildcards.sample_id} -X {params.max_reads} --readids \
            --dumpbase --skip-technical --gzip -Z &gt; {output}
        &quot;&quot;&quot;
</code></pre>
<p>Note that Snakemake now expects there to be a key named <code>max_reads</code> in the config 
dictionary. If we don't populate the dictionary somehow the dictionary will be 
empty so if you were to run the workflow now it would trigger a <code>KeyError</code> (try 
running <code>snakemake -s snakefile_mrsa.smk -n</code> to see for yourself). 
In order to populate the config dictionary with data for the workflow we could 
use the <code>snakemake --config KEY=VALUE</code> syntax directly from the command line.
However, from a reproducibility perspective, it's not optimal to set parameters 
from the command line, since it's difficult to keep track of which parameter 
values that were used. </p>
<p>A much better alternative is to use the <code>--configfile FILE</code> option to supply a 
configuration file to Snakemake. In this file we can collect all the 
project-specific settings, sample ids and so on. This also enables us to write 
the Snakefile in a more general manner so that it can be better reused between 
projects. Like several other files used in these tutorials, this file should be 
in <a href="https://en.wikipedia.org/wiki/YAML">yaml format</a>. Create the file below and 
save it as <code>config.yml</code>.</p>
<pre><code class="language-yaml">max_reads: 25000
</code></pre>
<p>If we now run Snakemake with <code>--configfile config.yml</code>, it will parse this file
to form the <code>config</code> dictionary. If you want to overwrite a parameter value,
<em>e.g.</em> for testing, you can still use the <code>--config KEY=VALUE</code> flag, as in 
<code>--config max_reads=1000</code>.</p>
<blockquote>
<p><strong>Tip</strong> <br>
Rather than supplying the config file from the command line you could also
add the line <code>configfile: "config.yml"</code> to the top of your Snakefile. Keep in 
mind that with such a setup Snakemake will complain if the file <code>config.yml</code> 
is not present.</p>
<p><strong>Quick recap</strong> <br>
In this section we've learned:</p>
<ul>
<li>How to set parameter values with the <code>params</code> directive.</li>
<li>How to run Snakemake with the <code>config</code> variable and with a configuration file.</li>
</ul>
</blockquote>

			<br>
			






<div class="row pt-2 pb-4">
	<div class="col-sm text-center text-sm-left">
		<a href="../snakemake-4-the-mrsa-workflow/" class="btn btn-sm btn-outline-dark">
			<i class="fa fa-chevron-left" aria-hidden="true"></i>
			Previous
		</a>
		<a href="../snakemake-4-the-mrsa-workflow/" class="btn btn-sm btn-link">
			The MRSA Workflow
		</a>
	</div>
	

	
	<div class="col-sm text-center text-sm-right">
		<a href="../snakemake-6-logs/" class="btn btn-sm btn-link">
			Logs
		</a>
		<a href="../snakemake-6-logs/" class="btn btn-sm btn-outline-dark">
			Next
			<i class="fa fa-chevron-right" aria-hidden="true"></i>
		</a>
	</div>
	
</div>

			<br>
		</div>

		<footer class="container-fluid wm-page-content text-center small">
			<p>
			<a href="https://github.com/Juke34/workshop-reproducible-research/edit/master/docs/snakemake/snakemake-5-parameters.md" target="_blank">Edit on workshop-reproducible-research</a>
			</p>
			<details>
				<summary>
					Documentation built with <a href="http://www.mkdocs.org/" target="_blank">MkDocs</a> using  <a href="https://github.com/Siphalor/mkdocs-custommill" target="_blank">CustomMill</a>.
				</summary>
				<h6 class="mt-2">Additional Licenses</h6>
				<ul><li>
						CustomMill is based on the <a href="https://github.com/gristlabs/mkdocs-windmill" target="_blank">Windmill</a> theme by Grist Labs, licensed under the <a href="https://github.com/gristlabs/mkdocs-windmill/blob/master/LICENSE" target="_blank">MIT License</a>
					</li>
					<li>
						CustomMill includes <a href="https://getbootstrap.com" target="_blank">Bootstrap</a> version 5.1.3. This is available under the <a href="https://github.com/twbs/bootstrap/blob/master/LICENSE" target="_blank">MIT License</a>.
					</li>
					<li>
						Syntax highlighting is implemented via <a href="https://highlightjs.org" target="_blank">highlight.js</a> licensed under the BSD 3-Clause License and the included github styling made by Vasily Polovnyov.
					</li>
					<li>
						Icons on this site are from <a href="https://fontawesome.com/" target="_blank">Font Awesome</a> which is licensed under the <a href="https://fontawesome.com/license/free" target="_blank">Font Awesome Free License</a>.
					</li></ul>
				<p>Built at 2023-02-10 16:41:13 UTC</p>
			</details>
		</footer>

		
	</body>
</html>