<!DOCTYPE html>
<html lang="en">
	<head>
		
		
		<meta charset="utf-8">
		<meta http-equiv="X-UA-Compatible" content="IE=edge">
		<meta name="viewport" content="width=device-width, initial-scale=1.0">
		
		<meta name="author" content="Jacques Dainat et al.">
		<link rel="canonical" href="https://Juke34.github.io/workshop-reproducible-research/jupyter/jupyter-8-the-mrsa-case-study/">
		<link rel="shortcut icon" href="../../img/favicon.ico">
		<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" />
		<title>The MRSA case study - Bioinformatics Workshop on Tools for Reproducible Research</title>
		<link href="../../bootstrap/dist/css/bootstrap.min.css" rel="stylesheet">
		<link href="../../css/font-awesome-5.12.2-all.min.css" rel="stylesheet">
		<link href="../../css/base.css" rel="stylesheet">
		<link rel="stylesheet" href="../../css/highlight-github.css">
		<script src="../../bootstrap/dist/js/bootstrap.min.js"></script>
		<script src="../../js/highlight.pack.js"></script>
		
		<script>
			var base_url = '../..';
			var home_url = '../../usage';
			var is_outer_page = false;
			
			var pageToc = [
				{title: "Sharing your work", url: "#_top", children: [
				]},
				{title: "Shared interactive notebooks", url: "#shared-interactive-notebooks", children: [
				]},
			];

		</script>
		<script src="../../js/base.js"></script> 
	</head>

	<body class="inner-page">
		

		<div class="container-fluid wm-page-content">
			<a name="_top" aria-hidden="true"></a>
			






<div class="row pt-2 pb-4">
	<div class="col-sm text-center text-sm-left">
		<a href="../jupyter-7-converting-notebooks/" class="btn btn-sm btn-outline-dark">
			<i class="fa fa-chevron-left" aria-hidden="true"></i>
			Previous
		</a>
		<a href="../jupyter-7-converting-notebooks/" class="btn btn-sm btn-link">
			Converting notebooks
		</a>
	</div>
	

	
	<div class="col-sm text-center text-sm-right">
		<a href="../jupyter-9-extra-material/" class="btn btn-sm btn-link">
			Extra material
		</a>
		<a href="../jupyter-9-extra-material/" class="btn btn-sm btn-outline-dark">
			Next
			<i class="fa fa-chevron-right" aria-hidden="true"></i>
		</a>
	</div>
	
</div>

			

			<p>As you might remember from the <a href="introduction">intro</a>, we are attempting to
understand how lytic bacteriophages can be used as a future therapy for the
multiresistant bacteria MRSA (methicillin-resistant <em>Staphylococcus aureus</em>).
We have already seen how to define the project environment in the <a href="conda-3-projects">Conda
tutorial</a> and how to set up the workflow in the <a href="snakemake-10-generalizing-workflows">Snakemake
tutorial</a>. Here we explore the results
from the Snakemake tutorial and generate a Supplementary Material file with
some basic stats.</p>
<p>In the <code>workshop-reproducible-research/tutorials/jupyter/</code> directory you will 
find a notebook called <code>supplementary_material.ipynb</code>. Open this notebook with 
Jupyter by running:</p>
<pre><code class="language-bash">jupyter notebook supplementary_material.ipynb
</code></pre>
<blockquote>
<p><strong>Tip</strong> <br>
Using what you've learned about markdown in notebooks, add headers 
and descriptive text to subdivide sections as you add them. This will
help you train how to structure and keep note of your work with a 
notebook.</p>
</blockquote>
<p>You will see that the notebook contains only a little markdown text and a code
cell with a function <code>get_geodata</code>. We'll start by adding a cell with some 
import statements. </p>
<p>Run the cell with the <code>get_geodata</code> function and add a new cell directly after 
it. Then add the following to the new cell:</p>
<pre><code class="language-python">import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
</code></pre>
<p>This imports the <code>pandas</code> (for working with tables), <code>seaborn</code> and 
<code>matplotlib.pyplot</code> (for plotting) and <code>numpy</code> (for numerical operations)
Python modules.</p>
<p>Also add:</p>
<pre><code class="language-python">import matplotlib_inline
matplotlib_inline.backend_inline.set_matplotlib_formats('pdf', 'svg')
</code></pre>
<p>to set high-quality output for plots.</p>
<p>Run the cell and create a new one below it.</p>
<p>In the next cell we'll define some parameters to use for the notebook:</p>
<pre><code class="language-python">counts_file=&quot;results/tables/counts.tsv&quot;
summary_file=&quot;results/tables/counts.tsv.summary&quot;
multiqc_file=&quot;intermediate/multiqc_general_stats.txt&quot;
rulegraph_file=&quot;results/rulegraph.png&quot;
SRR_IDs=[&quot;SRR935090&quot;,&quot;SRR935091&quot;,&quot;SRR935092&quot;]
GSM_IDs=[&quot;GSM1186459&quot;,&quot;GSM1186460&quot;,&quot;GSM1186461&quot;]
GEO_ID=&quot;GSE48896&quot;
</code></pre>
<p>As you can see we add paths to results files and define lists with some sample 
IDS. Run this cell and add a new one below it.</p>
<p>Next, we'll fetch some sample information from NCBI using the <code>get_geodata</code>
function defined at the start of the notebook and collate it into a dataframe.</p>
<pre><code class="language-python">id_df = pd.DataFrame(data=GSM_IDs, index=SRR_IDs, columns=[&quot;geo_accession&quot;])
geo_df = get_geodata(GEO_ID)
name_df = pd.merge(id_df, geo_df, left_on=&quot;geo_accession&quot;, right_index=True)
# Create a dictionary to rename sample ids in downstream plots
name_dict = name_df.to_dict()
</code></pre>
<p>Run the cell and take a look at the contents of the <code>name_df</code> dataframe (<em>e.g.</em>
run a cell with that variable only to output it below the cell).</p>
<p>Now we'll load some statistics from the QC part of the workflow, specifically
the 'general_stats' file from <code>multiqc</code>. Add the following to a new cell and run
it:</p>
<pre><code class="language-python">qc = pd.read_csv(multiqc_file, sep=&quot;\t&quot;)
qc.rename(columns=lambda x: x.replace(&quot;FastQC_mqc-generalstats-fastqc-&quot;, &quot;&quot;).replace(&quot;_&quot;, &quot; &quot;), inplace=True)
qc = pd.merge(qc, name_df, left_on=&quot;Sample&quot;, right_index=True)
qc
</code></pre>
<p>In the code above we load the multiqc file, rename the columns by stripping the 
<code>FastQC_mqc-generalstats-fastqc-</code> part from column names and replace underscores
with spaces. Finally the table is merged with the information obtained in the 
step above and output to show summary statistics from the QC stage.</p>
<p>Next it's time to start loading gene count results from the workflow. Start by 
reading the counts and summary results, then edit the columns and index:</p>
<pre><code class="language-python"># Read count data
counts = pd.read_csv(counts_file, sep=&quot;\t&quot;, header=0, comment=&quot;#&quot;, index_col=0)
# Read summary data
counts_summary = pd.read_csv(summary_file, sep=&quot;\t&quot;, index_col=0)
# Rename columns to extract SRR ids
counts.rename(columns = lambda x: x.split(&quot;/&quot;)[-1].replace(&quot;.sorted.bam&quot;,&quot;&quot;), inplace=True)
counts_summary.rename(columns = lambda x: x.split(&quot;/&quot;)[-1].replace(&quot;.sorted.bam&quot;,&quot;&quot;), inplace=True)
</code></pre>
<p>Take a look at the <code>counts</code> dataframe to get an idea of the data structure. As
you can see the dataframe shows genes as rows while the columns shows various
information such as start and stop, strand and length of the genes. The last
three columns contain counts of the genes in each of the samples. </p>
<p>If you have a look at the <code>counts_summary</code> dataframe you will see statistics 
from the read assignment step, showing number of reads that could be properly 
assigned as well as number of reads that could not be assigned to genes for 
various reasons.</p>
<p>Now let's generate a barplot of the summary statistics. Before we plot, we'll
remove rows that have only zero values:</p>
<pre><code class="language-python"># Remove rows with only zero values
summary_plot_data = counts_summary.loc[counts_summary.sum(axis=1)&gt;0]
</code></pre>
<p>Now for the plotting:</p>
<pre><code class="language-python"># Set color palette to 'Set2'
colors = sns.color_palette(&quot;Set2&quot;)
# Create a stacked barplot
ax = summary_plot_data.T.plot(kind=&quot;bar&quot;, stacked=True, color=colors)
# Move legend and set legend title
ax.legend(bbox_to_anchor=(1,1), title=&quot;Category&quot;);
</code></pre>
<p>The final plot will be a heatmap of gene counts for a subset of the genes. We'll
select genes whose standard deviation/mean count across samples is greater than 
1.5, <strong>and</strong> have a maximum of at least 5 reads in 1 or more sample:</p>
<pre><code class="language-python"># Slice the dataframe to only sample counts
count_data = counts.loc[:, SRR_IDs]
# Filter to genes with std/mean &gt; 1.2 and with a max of at least 5
heatmap_data = count_data.loc[(count_data.std(axis=1).div(count_data.mean(axis=1))&gt;1.2)&amp;(count_data.max(axis=1)&gt;5)]
</code></pre>
<p>We'll also replace the SRR ids with the title of samples
used in the study, using the <code>name_dict</code> dictionary created further up in the 
notebook:</p>
<pre><code class="language-python">heatmap_data = heatmap_data.rename(columns = name_dict['title'])
</code></pre>
<p>Now let's plot the heatmap. We'll log-transform the counts, set color scale 
to Blue-Yellow-Red and cluster both samples and genes using 'complete' linkage
clustering:</p>
<pre><code class="language-python">with sns.plotting_context(&quot;notebook&quot;, font_scale=0.7):
    ax = sns.clustermap(data=np.log10(heatmap_data+1), cmap=&quot;RdYlBu_r&quot;, 
                        method=&quot;complete&quot;, yticklabels=True, linewidth=.5,
                        cbar_pos=(.7, .85, .05, .1), figsize=(3,9))
    plt.setp(ax.ax_heatmap.get_xticklabels(), rotation=270)
</code></pre>
<p>In the code above we use the seaborn <code>plotting_context</code> function to scale all 
text elements of the heatmap in one go.</p>
<p>As a final step we'll add some info for reproducibility under the 
<strong>Reproducibility</strong> section. To add the overview image of the workflow found in
<code>results/rulegraph.png</code> we can use the <code>Image</code> function from <code>IPython.display</code>:</p>
<pre><code class="language-python">from IPython.display import Image
Image(rulegraph_file)
</code></pre>
<p>Let's also output the full conda environment so that all packages and versions 
are included in the notebook. There are several ways this can be done, for 
example you could simply add:</p>
<pre><code class="language-python">!conda list
</code></pre>
<p>to the end of the notebook.</p>
<blockquote>
<p><strong>Tip</strong> <br>
If you want to know more about how notebooks can be integrated into 
Snakemake worfklows, see the Extra material at the end of this tutorial</p>
</blockquote>
<h2 id="sharing-your-work">Sharing your work<a class="headerlink" href="#sharing-your-work" title="Permanent link">#</a></h2>
<p>The files you're working with come from a GitHub repo. Both GitHub and Bitbucket
can render Jupyter notebooks as well as other types of Markdown documents. Now
go to our GitHub repo at
<a href="https://github.com/NBISweden/workshop-reproducible-research">https://github.com/NBISweden/workshop-reproducible-research</a>
and navigate to <code>tutorials/jupyter/supplementary_material.ipynb</code>.</p>
<p><img alt="" src="images/jupyter_supplementary.png" />{ width=700px }</p>
<p>As you can imagine, having this very effortless way of sharing results
can greatly increase the visibility of your work. You work as normal on
your project, and push regularly to the repository as you would anyways,
and the output is automatically available for anyone to see. Or for a
select few if you're not ready to share your findings with the world
quite yet.</p>
<p>Say your notebook isn't on Github/Bitbucket. All hope isn't lost there.
Jupyter.org provides a neat functionality called <em>nbviewer</em>, where you can
paste a URL to any notebook and they will render it for you. Go to
<a href="https://nbviewer.jupyter.org">https://nbviewer.jupyter.org</a> and try
this out with our notebook.</p>
<pre><code class="language-no-highlight">https://raw.githubusercontent.com/NBISweden/workshop-reproducible-research/main/tutorials/jupyter/supplementary_material.ipynb
</code></pre>
<h2 id="shared-interactive-notebooks">Shared interactive notebooks<a class="headerlink" href="#shared-interactive-notebooks" title="Permanent link">#</a></h2>
<p>So far we've only shared static representations of notebooks. A strong
trend at the moment is to run your notebooks in the cloud, so that the
person you want to share with could actually execute and modify your
code. This is a great way of increasing visibility and letting
collaborators or readers get more hands-on with your data and analyses.
From a reproducibility perspective, there are both advantages and
drawbacks. On the plus side is that running your work remotely forces
you to be strict when it comes to defining the environment it uses
(probably in the form of a Conda environment or Docker image). On the
negative side is that you become reliant on a third-party service that
might change input formats, go out of business, or change payment model.</p>
<p>Here we will try out a service called Binder, which lets you run and
share Jupyter Notebooks in Git repositories for free. There are a number
of <a href="https://github.com/binder-examples/">example repositories</a> that are
setup to be used with Binder. Navigate to
<a href="https://github.com/binder-examples/conda/">https://github.com/binder-examples/conda/</a> 
to see one such example. As you can see the repository contains a LICENSE 
file, a README, an environment file and a notebook. To use a repository 
with Binder the environment file should contain all the packages needed 
to run notebooks in the repo. So let's try to run the <code>index.ipynb</code> file 
using Binder:</p>
<p>Just go to <a href="https://mybinder.org">https://mybinder.org</a> and paste the link 
to the GitHub repo. Note the link that you can use to share your notebook. 
Then press "launch".</p>
<p><img alt="" src="images/binder.png" />{ width=700px }</p>
<p>What will happen now is that:</p>
<ul>
<li>Binder detects the <code>environment.yml</code> file in the root of the repo.
  Binder then builds a <em>Docker image</em> based on the file. This might take
  a minute or two. You can follow the progress in the build log.</li>
<li>Binder then launches the Jupyter Notebook server in the Docker
  container..</li>
<li>..and opens a browser tab with it for you.</li>
</ul>
<p>Once the process is finished you will be presented with a Jupyter server
overview of the contents in the repository. Click on the <code>index.ipynb</code>
notebook to open it. Tada! You are now able to interact with (and
modify) someone else's notebook online.</p>
<p>Applied to your own projects you now have a way to run analyses in the
cloud and in an environment that you define yourself. All that's needed
for someone to replicate your analyses is that you share a link with
them. Note that notebooks on Binder are read-only; its purpose is for
trying out and showing existing notebooks rather than making new ones.</p>
<blockquote>
<p><strong>Tip</strong> <br>
By default Binder looks for configuration files such as environment.yml
in the root of the repository being built. But you may also put 
such files outside the root by making a <code>binder/</code> folder in the root
and placing the file there.  </p>
<p><strong>A note on transparency</strong> <br>
Resources like Github/Bitbucket and Jupyter Notebooks have changed 
the way we do scientific research by encouraging visibility, social 
interaction and transparency. 
It was not long ago that the analysis scripts and workflows in a lab were
well-guarded secrets that we only most reluctantly shared with others.
Assuming that it was even possible. In most cases, the only postdoc who
knew how to get it to work had left for a new position in industry, or
no one could remember the password to the file server. If you're a PhD
student, we encourage you to embrace this new development
wholeheartedly, for it will make your research better and make you into
a better scientist. And you will have more fun.</p>
<p><strong>Quick recap</strong> <br>
In this section we've learned:</p>
<ul>
<li>How notebooks can be used to generate summary statistics and plots using the 
results of a workflow run</li>
<li>How to share notebooks via <a href="https://nbviewer.jupyter.org">nbviewer</a> and 
<a href="https://mybinder.org">Binder</a> </li>
</ul>
</blockquote>

			<br>
			






<div class="row pt-2 pb-4">
	<div class="col-sm text-center text-sm-left">
		<a href="../jupyter-7-converting-notebooks/" class="btn btn-sm btn-outline-dark">
			<i class="fa fa-chevron-left" aria-hidden="true"></i>
			Previous
		</a>
		<a href="../jupyter-7-converting-notebooks/" class="btn btn-sm btn-link">
			Converting notebooks
		</a>
	</div>
	

	
	<div class="col-sm text-center text-sm-right">
		<a href="../jupyter-9-extra-material/" class="btn btn-sm btn-link">
			Extra material
		</a>
		<a href="../jupyter-9-extra-material/" class="btn btn-sm btn-outline-dark">
			Next
			<i class="fa fa-chevron-right" aria-hidden="true"></i>
		</a>
	</div>
	
</div>

			<br>
		</div>

		<footer class="container-fluid wm-page-content text-center small">
			<p>
			<a href="https://github.com/Juke34/workshop-reproducible-research/edit/master/docs/jupyter/jupyter-8-the-mrsa-case-study.md" target="_blank">Edit on workshop-reproducible-research</a>
			</p>
			<details>
				<summary>
					Documentation built with <a href="http://www.mkdocs.org/" target="_blank">MkDocs</a> using  <a href="https://github.com/Siphalor/mkdocs-custommill" target="_blank">CustomMill</a>.
				</summary>
				<h6 class="mt-2">Additional Licenses</h6>
				<ul><li>
						CustomMill is based on the <a href="https://github.com/gristlabs/mkdocs-windmill" target="_blank">Windmill</a> theme by Grist Labs, licensed under the <a href="https://github.com/gristlabs/mkdocs-windmill/blob/master/LICENSE" target="_blank">MIT License</a>
					</li>
					<li>
						CustomMill includes <a href="https://getbootstrap.com" target="_blank">Bootstrap</a> version 5.1.3. This is available under the <a href="https://github.com/twbs/bootstrap/blob/master/LICENSE" target="_blank">MIT License</a>.
					</li>
					<li>
						Syntax highlighting is implemented via <a href="https://highlightjs.org" target="_blank">highlight.js</a> licensed under the BSD 3-Clause License and the included github styling made by Vasily Polovnyov.
					</li>
					<li>
						Icons on this site are from <a href="https://fontawesome.com/" target="_blank">Font Awesome</a> which is licensed under the <a href="https://fontawesome.com/license/free" target="_blank">Font Awesome Free License</a>.
					</li></ul>
				<p>Built at 2023-02-10 16:41:13 UTC</p>
			</details>
		</footer>

		
	</body>
</html>