<!DOCTYPE html>
<html lang="en">
	<head>
		
		
		<meta charset="utf-8">
		<meta http-equiv="X-UA-Compatible" content="IE=edge">
		<meta name="viewport" content="width=device-width, initial-scale=1.0">
		
		<meta name="author" content="Jacques Dainat et al.">
		<link rel="canonical" href="https://Juke34.github.io/workshop-reproducible-research/rmarkdown/r-markdown-6-the-mrsa-case-study/">
		<link rel="shortcut icon" href="../../img/favicon.ico">
		<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" />
		<title>The MRSA case study - Bioinformatics Workshop on Tools for Reproducible Research</title>
		<link href="../../bootstrap/dist/css/bootstrap.min.css" rel="stylesheet">
		<link href="../../css/font-awesome-5.12.2-all.min.css" rel="stylesheet">
		<link href="../../css/base.css" rel="stylesheet">
		<link rel="stylesheet" href="../../css/highlight-github.css">
		<script src="../../bootstrap/dist/js/bootstrap.min.js"></script>
		<script src="../../js/highlight.pack.js"></script>
		
		<script>
			var base_url = '../..';
			var home_url = '../../usage';
			var is_outer_page = false;
			
			var pageToc = [
				{title: "Overview", url: "#_top", children: [
				]},
				{title: "Rendering options and paths", url: "#rendering-options-and-paths", children: [
				]},
				{title: "Formatting tables and figures", url: "#formatting-tables-and-figures", children: [
				]},
			];

		</script>
		<script src="../../js/base.js"></script> 
	</head>

	<body class="inner-page">
		

		<div class="container-fluid wm-page-content">
			<a name="_top" aria-hidden="true"></a>
			






<div class="row pt-2 pb-4">
	<div class="col-sm text-center text-sm-left">
		<a href="../r-markdown-5-rendering/" class="btn btn-sm btn-outline-dark">
			<i class="fa fa-chevron-left" aria-hidden="true"></i>
			Previous
		</a>
		<a href="../r-markdown-5-rendering/" class="btn btn-sm btn-link">
			Rendering
		</a>
	</div>
	

	
	<div class="col-sm text-center text-sm-right">
		<a href="../r-markdown-7-extra-material/" class="btn btn-sm btn-link">
			Extra materiel
		</a>
		<a href="../r-markdown-7-extra-material/" class="btn btn-sm btn-outline-dark">
			Next
			<i class="fa fa-chevron-right" aria-hidden="true"></i>
		</a>
	</div>
	
</div>

			

			<p>As you might remember from the <a href="introduction">intro</a>, we are attempting to
understand how lytic bacteriophages can be used as a future therapy for the
multiresistant bacteria MRSA (methicillin-resistant <em>Staphylococcus aureus</em>).
In this exercise we will use R Markdown to make a report in form of
a Supplementary Material HTML based on the outputs from the <a href="snakemake-10-generalizing-workflows">Snakemake
tutorial</a>. Among the benefits of having
the supplementary material (or even the full manuscript) in R Markdown format
are:</p>
<ul>
<li>It is fully transparent how the text, tables and figures were produced.</li>
<li>If you get reviewer comments, or realize you've made a mistake somewhere, you
  can easily update your code and regenerate the document with the push of
  a button.</li>
<li>By making report generation part of your workflow early on in a project, much
  of the eventual manuscript "writes itself". You no longer first have to
  finish the research part and <em>then</em> start creating the tables and figures for
  the paper.</li>
</ul>
<p>Before you start:</p>
<ul>
<li>Make sure that your working directory in R is <code>workshop-reproducible-research/tutorials/rmarkdown</code> 
  in the course directory (Session &gt; Set Working Directory).</li>
<li>Open the file <code>code/supplementary_material.Rmd</code>.</li>
</ul>
<blockquote>
<p><strong>Note</strong> <br>
In this tutorial we have used Conda to install all the R packages we need,
so that you get to practice how you can actually do this in projects of your
own. You can, however, install things using <code>install.packages()</code> or
<code>BiocManager::install()</code> as well, even though this makes it both less
reproducible and more complicated in most cases.</p>
</blockquote>
<h2 id="overview">Overview<a class="headerlink" href="#overview" title="Permanent link">#</a></h2>
<p>Let's start by taking a look at the YAML header at the top of the file. The
parameters correspond to files (and sample IDs) that are generated by the MRSA
analysis workflow (see the <a href="snakemake-10-generalizing-workflows">Snakemake tutorial</a>)
and contain results that we want to include in the supplementary material
document. We've also specified that we want to render to HTML.</p>
<pre><code class="language-yaml">---
title: &quot;Supplementary Materials&quot;
output: html_document
params:
    counts_file: &quot;results/tables/counts.tsv&quot;
    multiqc_file: &quot;intermediate/multiqc_general_stats.txt&quot;
    summary_file: &quot;results/tables/counts.tsv.summary&quot;
    rulegraph_file: &quot;results/rulegraph.png&quot;
    SRR_IDs: &quot;SRR935090 SRR935091 SRR935092&quot;
    GSM_IDs: &quot;GSM1186459 GSM1186460 GSM1186461&quot;
---
</code></pre>
<ul>
<li>From a reproducibility perspective it definitely makes sense to include
  information about who authored the document and the date it was generated.
  Add the two lines below to the YAML header. Note that we can include inline
  R code by using <code>`r some_code`</code>.</li>
</ul>
<pre><code>author: John Doe, Joan Dough, Jan Doh, Dyon Do
date: &quot;`r format(Sys.time(), '%d %B, %Y')`&quot;
</code></pre>
<blockquote>
<p><strong>Tip</strong> <br>
Make it a practice to keep track of all input files and add them as
parameters rather than hard-coding them later in the R code.</p>
</blockquote>
<p>Next, take a look at the <code>dependencies</code>, <code>read_params</code>, and <code>read_data</code> chunks.
They 1) load the required packages, 2) read the parameters and store them in
R objects to be used later in the code, and 3) read the data in the counts
file, the multiqc file, as well as fetch meta data from GEO. These chunks are
provided as is, and you do not need to edit them.</p>
<p>Below these chunks there is some markdown text that contains the Supplementary
Methods section. Note the use of section headers using <code>#</code> and <code>##</code>. Then there
is a Supplementary Tables and Figures section. This contains four code chunks,
each for a specific table or figure. Have a quick look at the code and see if
you can figure out what it does, but don't worry if you can't understand
everything.</p>
<p>Finally, there is a <em>Reproducibility</em> section which describes how the results in
the report can be reproduced. The <code>session_info</code> chunk prints information
regarding R version and which packages and versions that are used. We highly
encourage you to include this chunk in all your R Markdown reports: it's an
effortless way to increase reproducibility.</p>
<h2 id="rendering-options-and-paths">Rendering options and paths<a class="headerlink" href="#rendering-options-and-paths" title="Permanent link">#</a></h2>
<ul>
<li>Now that you have had a look at the R Markdown document, it is time to Knit!
  We will do this from the R terminal (rather than pressing <em>Knit</em>).</li>
</ul>
<pre><code class="language-r">rmarkdown::render(&quot;code/supplementary_material.Rmd&quot;, output_dir = &quot;results&quot;)
</code></pre>
<p>The reason for this is that we can then redirect the output html file to be
saved in the <code>results/</code> directory.</p>
<p>Normally, while rendering, R code in the Rmd file will be executed using the
directory of the Rmd file as working directory (<code>rmarkdown/code</code> in this case).
However, it is good practice to write all code as if it would be executed from
the project root directory (<code>rmarkdown/</code> in this case). For instance, you can
see that we have specified the files in <code>params</code> with relative paths from the
project root directory. To set a different directory as working directory for
all chunks one modifies the knit options like this:</p>
<pre><code class="language-r">knitr::opts_knit$set(root.dir = '../')
</code></pre>
<p>Here we set the working directory to the parent directory of the Rmd file
(<code>../</code>), in other words, the project root. Use this rather than <code>setwd()</code> while
working with Rmd files.</p>
<ul>
<li>Take a look at the output. You should find the html file in the <code>results</code>
  directory.</li>
</ul>
<h2 id="formatting-tables-and-figures">Formatting tables and figures<a class="headerlink" href="#formatting-tables-and-figures" title="Permanent link">#</a></h2>
<p>You will probably get a good idea of the contents of the file, but the tables
look weird and the figures could be better formatted. Let's start by adjusting
the figures!</p>
<ul>
<li>
<p>Locate the <code>Setup</code> chunk. Here, we have already set <code>echo = FALSE</code>. Let's add
  some default figure options: <code>fig.height = 6, fig.width = 6, fig.align
  = 'center'</code>. This will make the figures slightly smaller than default and
  center them.</p>
</li>
<li>
<p>Knit again, using the same R command as above. Do you notice any difference?
  Better, but still not perfect!</p>
</li>
</ul>
<p>Let's improve the tables! We have not talked about tables before. There are
several options to print tables, here we will use the <code>kable</code> function which is
part of the <code>knitr</code> package.</p>
<ul>
<li>Go to the <code>Sample info</code> chunk. Replace the last line, <code>sample_info</code>, with:</li>
</ul>
<pre><code class="language-r">knitr::kable(sample_info)
</code></pre>
<ul>
<li>Knit again and look at the result. You should see a formatted table.</li>
<li>The column names can be improved, and we could use a table legend. Change to
  use the following:</li>
</ul>
<pre><code class="language-r">knitr::kable(sample_info, caption = &quot;Sample info&quot;,
             col.names = c(&quot;SRR&quot;, &quot;GEO&quot;, &quot;Strain&quot;, &quot;Treatment&quot;))
</code></pre>
<ul>
<li>Knit and check the result.</li>
<li>Try to fix the table in the <code>QC statistics</code> chunk in the same manner. The
  column names are fine here so no need to change them, but add a table legend:
  "QC stats from FastQC". Knit and check your results.</li>
</ul>
<p>Let's move on to the figures!</p>
<ul>
<li>Go to the <code>Counts barplot</code> chunk. To add a figure legend we have to use
  a chunk option (so not in the same way as for tables). Add the chunk option:</li>
</ul>
<pre><code class="language-r">fig.cap = &quot;Counting statistics per sample, in terms of read counts for genes
           and reads not counted for various reasons.&quot;
</code></pre>
<ul>
<li>Knit and check the outcome!</li>
<li>Next, add a figure legend to the figure in the <code>gene-heatmap</code> chunk. Here we
  can try out the possibility to add R code to generate the legend:</li>
</ul>
<pre><code class="language-r">fig.cap = paste0(&quot;Expression (log-10 counts) of genes with at least &quot;,
                 max_cutoff, &quot; counts in one sample and a CV&gt;&quot;, cv_cutoff, &quot;.&quot;)
</code></pre>
<p>This will use the <code>cv_cutoff</code> and <code>max_cutoff</code> variables to ensure that the
figure legend gives the same information as was used to generate the plot. Note
that figure legends are generated <em>after</em> the corresponding code chunk is
evaluated. This means we can use objects defined in the code chunk in the
legend.</p>
<ul>
<li>Knit and have a look at the results.</li>
</ul>
<p>The heatmap still looks a bit odd. Let's play with the <code>fig.height</code> and
<code>out.height</code> options, like we did above, to scale the figure in a more
appropriate way. Add this to the chunk options: <code>fig.height = 10,
out.height = "22cm"</code>. Knit and check the results. Does it look better now?</p>
<ul>
<li>Now let's add a third figure! This time we will not plot a figure in R, but
  use an available image file showing the structure of the Snakemake workflow
  used to generate the inputs for this report. Add a new chunk at the end of
  the Supplementary Tables and Figures section containing this code:</li>
</ul>
<pre><code class="language-r">knitr::include_graphics(normalizePath(rulegraph_file))
</code></pre>
<blockquote>
<p><strong>Knitr paths</strong> <br>
Just like for R Markdown paths in general, Knitr needs everything to be
relative to the directory in which the <code>.Rmd</code> file is situated. Just like
setting the <code>root.dir</code> chunk option can help with this for the overall
rendering, the <code>normalizePath()</code> function is needed when using
<code>include_graphics()</code> for adding images with Knitr.</p>
</blockquote>
<ul>
<li>Also, add the chunk options:</li>
</ul>
<pre><code class="language-r">fig.cap = &quot;A rule graph showing the different steps of the bioinformatic
           analysis that is included in the Snakemake workflow.&quot;
</code></pre>
<p>and:</p>
<pre><code class="language-r">out.height = &quot;11cm&quot;
</code></pre>
<ul>
<li>Knit and check the results.</li>
</ul>
<blockquote>
<p><strong>Note</strong> <br>
It is definitely possible to render R Markdown documents as part of
a Snakemake or Nextflow workflow. This is something we do for the final
version of the MRSA project (in the Containers tutorial). In such cases it is
advisable to manage the installation of R and required R packages through
your conda environment file and use the <code>rmarkdown::render()</code> command from
the shell section of your Snakemake rule or Nexflow process.</p>
<p><strong>Quick recap</strong> <br>
In this section you learned some additional details for making nice
R Markdown reports in a reproducible research project setting, including
setting the root directory, adding tables as well as setting figure and
table captions.</p>
</blockquote>

			<br>
			






<div class="row pt-2 pb-4">
	<div class="col-sm text-center text-sm-left">
		<a href="../r-markdown-5-rendering/" class="btn btn-sm btn-outline-dark">
			<i class="fa fa-chevron-left" aria-hidden="true"></i>
			Previous
		</a>
		<a href="../r-markdown-5-rendering/" class="btn btn-sm btn-link">
			Rendering
		</a>
	</div>
	

	
	<div class="col-sm text-center text-sm-right">
		<a href="../r-markdown-7-extra-material/" class="btn btn-sm btn-link">
			Extra materiel
		</a>
		<a href="../r-markdown-7-extra-material/" class="btn btn-sm btn-outline-dark">
			Next
			<i class="fa fa-chevron-right" aria-hidden="true"></i>
		</a>
	</div>
	
</div>

			<br>
		</div>

		<footer class="container-fluid wm-page-content text-center small">
			<p>
			<a href="https://github.com/Juke34/workshop-reproducible-research/edit/master/docs/rmarkdown/r-markdown-6-the-mrsa-case-study.md" target="_blank">Edit on workshop-reproducible-research</a>
			</p>
			<details>
				<summary>
					Documentation built with <a href="http://www.mkdocs.org/" target="_blank">MkDocs</a> using  <a href="https://github.com/Siphalor/mkdocs-custommill" target="_blank">CustomMill</a>.
				</summary>
				<h6 class="mt-2">Additional Licenses</h6>
				<ul><li>
						CustomMill is based on the <a href="https://github.com/gristlabs/mkdocs-windmill" target="_blank">Windmill</a> theme by Grist Labs, licensed under the <a href="https://github.com/gristlabs/mkdocs-windmill/blob/master/LICENSE" target="_blank">MIT License</a>
					</li>
					<li>
						CustomMill includes <a href="https://getbootstrap.com" target="_blank">Bootstrap</a> version 5.1.3. This is available under the <a href="https://github.com/twbs/bootstrap/blob/master/LICENSE" target="_blank">MIT License</a>.
					</li>
					<li>
						Syntax highlighting is implemented via <a href="https://highlightjs.org" target="_blank">highlight.js</a> licensed under the BSD 3-Clause License and the included github styling made by Vasily Polovnyov.
					</li>
					<li>
						Icons on this site are from <a href="https://fontawesome.com/" target="_blank">Font Awesome</a> which is licensed under the <a href="https://fontawesome.com/license/free" target="_blank">Font Awesome Free License</a>.
					</li></ul>
				<p>Built at 2023-02-10 16:41:13 UTC</p>
			</details>
		</footer>

		
	</body>
</html>