<!DOCTYPE html>
<html lang="en">
	<head>
		
		
		<meta charset="utf-8">
		<meta http-equiv="X-UA-Compatible" content="IE=edge">
		<meta name="viewport" content="width=device-width, initial-scale=1.0">
		
		<meta name="author" content="Jacques Dainat et al.">
		<link rel="canonical" href="https://Juke34.github.io/workshop-reproducible-research/introduction/introduction/">
		<link rel="shortcut icon" href="../../img/favicon.ico">
		<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" />
		<title>Introduction - Bioinformatics Workshop on Tools for Reproducible Research</title>
		<link href="../../bootstrap/dist/css/bootstrap.min.css" rel="stylesheet">
		<link href="../../css/font-awesome-5.12.2-all.min.css" rel="stylesheet">
		<link href="../../css/base.css" rel="stylesheet">
		<link rel="stylesheet" href="../../css/highlight-github.css">
		<script src="../../bootstrap/dist/js/bootstrap.min.js"></script>
		<script src="../../js/highlight.pack.js"></script>
		
		<script>
			var base_url = '../..';
			var home_url = '../../usage';
			var is_outer_page = false;
			
			var pageToc = [
				{title: "The case study", url: "#_top", children: [
						
				{title: "Background", url: "#_top", children: [
				]},
				{title: "Analysis", url: "#analysis", children: [
				]},
				]},
			];

		</script>
		<script src="../../js/base.js"></script> 
	</head>

	<body class="inner-page">
		

		<div class="container-fluid wm-page-content">
			<a name="_top" aria-hidden="true"></a>
			






<div class="row pt-2 pb-4">
	<div class="col-sm text-center text-sm-left">
		<a href="../../course-information/pre-course-setup/" class="btn btn-sm btn-outline-dark">
			<i class="fa fa-chevron-left" aria-hidden="true"></i>
			Previous
		</a>
		<a href="../../course-information/pre-course-setup/" class="btn btn-sm btn-link">
			Setup
		</a>
	</div>
	

	
	<div class="col-sm text-center text-sm-right">
		<a href="../../git/git-1-introduction/" class="btn btn-sm btn-link">
			Introduction
		</a>
		<a href="../../git/git-1-introduction/" class="btn btn-sm btn-outline-dark">
			Next
			<i class="fa fa-chevron-right" aria-hidden="true"></i>
		</a>
	</div>
	
</div>

			

			<p>Welcome to the tutorials! Here we will learn how to make a computational
research project reproducible using several different tools, described in the
figure below:</p>
<p><img alt="" src="images/reproducibility_overview_with_logos.png" />{ width=700px }</p>
<p>The figure above gives an overview of the different parts of computational
reproducibility (data, code, workflow and environment), as well as the various
tools that are used for each part; Git is, arguably, integral to all of the
parts, but we only listed it in the <em>code</em> section for a less cluttered figure.</p>
<p>The course has a tutorial for each of the tools, all made so that they can be
completed independently of each other. It is therefore perfectly
possible to go through them in whatever order you prefer, but we suggest the
following order:</p>
<ol>
<li>Git</li>
<li>Conda</li>
<li>Snakemake</li>
<li>Nextflow</li>
<li>R Markdown</li>
<li>Jupyter</li>
<li>Containers</li>
</ol>
<p>You will find the tutorials in the <em>Modules</em> section in the navigation menu.</p>
<p>Please make sure to carefully follow the <a href="https://uppsala.instructure.com/courses/COURSE_ID/pages/pre-course-setup">pre-course setup</a> 
to install the tools and download the course material before starting with any of the 
tutorials. These will create quite a lot of files on your computer, some of which 
will actually take up a bit of storage space too. In order to remove any traces of 
these after completing the tutorials, please refer to the 
<a href="https://uppsala.instructure.com/courses/COURSE_ID/pages/take-down">Take down section</a>.</p>
<p>Before going into the tutorials themselves, we first describe the case study
from which the example data comes from. </p>
<h1 id="the-case-study">The case study<a class="headerlink" href="#the-case-study" title="Permanent link">#</a></h1>
<p>We will be running a small bioinformatics project as a case study, and use that
to exemplify the different steps of setting up a reproducible research project.
To give you some context, the study background and analysis steps are briefly
described below.</p>
<h4 id="background">Background<a class="headerlink" href="#background" title="Permanent link">#</a></h4>
<p>The data is taken from <a href="http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0076572">Osmundson, Dewell, and Darst (2013)</a>,
who have studied methicillin-resistant <em>Staphylococcus aureus</em> (MRSA). MRSA is
resistant to broad spectrum beta-lactam antibiotics and lead to
difficult-to-treat infections in humans. Lytic bacteriophages have been
suggested as potential therapeutic agents, or as the source of novel antibiotic
proteins or peptides. One such protein, gp67, was identified as
a transcription-inhibiting transcription factor with an antimicrobial effect.
To identify <em>S. aureus</em> genes repressed by gp67, the authors expressed gp67 in
<em>S. aureus</em> cells. RNA-seq was then performed on three S. aureus strains:</p>
<ul>
<li>RN4220 with pRMC2 with gp67</li>
<li>RN4220 with empty pRMC2</li>
<li>NCTC8325-4</li>
</ul>
<h4 id="analysis">Analysis<a class="headerlink" href="#analysis" title="Permanent link">#</a></h4>
<p>The graph below shows the different steps of the analysis that are included in
this project:</p>
<p><img alt="" src="images/mrsa-overview.png" />{ width=400px }</p>
<p>The input files are:</p>
<ul>
<li>RNA-seq raw data (FASTQ files) for the three strains</li>
<li><em>S. aureus</em> genome sequence (a FASTA file)</li>
<li><em>S. aureus</em> genome annotation (a GFF file)</li>
</ul>
<p>The workflow itself will perform the following tasks:</p>
<ul>
<li>Downloading and indexing of the reference genome using <em>Bowtie2</em></li>
<li>Downloading the raw FASTQ data from the <em>Sequence Read Archive</em> (SRA)</li>
<li>Run some quality controls on the data using <em>FastQC</em> and <em>MultiQC</em></li>
<li>Align the raw data to the genome and calculate the gene expression using
  <em>featureCounts</em></li>
<li>Produce supplementary materials using data from quality controls, gene
  expression and the workflow figure shown above</li>
</ul>

			<br>
			






<div class="row pt-2 pb-4">
	<div class="col-sm text-center text-sm-left">
		<a href="../../course-information/pre-course-setup/" class="btn btn-sm btn-outline-dark">
			<i class="fa fa-chevron-left" aria-hidden="true"></i>
			Previous
		</a>
		<a href="../../course-information/pre-course-setup/" class="btn btn-sm btn-link">
			Setup
		</a>
	</div>
	

	
	<div class="col-sm text-center text-sm-right">
		<a href="../../git/git-1-introduction/" class="btn btn-sm btn-link">
			Introduction
		</a>
		<a href="../../git/git-1-introduction/" class="btn btn-sm btn-outline-dark">
			Next
			<i class="fa fa-chevron-right" aria-hidden="true"></i>
		</a>
	</div>
	
</div>

			<br>
		</div>

		<footer class="container-fluid wm-page-content text-center small">
			<p>
			<a href="https://github.com/Juke34/workshop-reproducible-research/edit/master/docs/introduction/introduction.md" target="_blank">Edit on workshop-reproducible-research</a>
			</p>
			<details>
				<summary>
					Documentation built with <a href="http://www.mkdocs.org/" target="_blank">MkDocs</a> using  <a href="https://github.com/Siphalor/mkdocs-custommill" target="_blank">CustomMill</a>.
				</summary>
				<h6 class="mt-2">Additional Licenses</h6>
				<ul><li>
						CustomMill is based on the <a href="https://github.com/gristlabs/mkdocs-windmill" target="_blank">Windmill</a> theme by Grist Labs, licensed under the <a href="https://github.com/gristlabs/mkdocs-windmill/blob/master/LICENSE" target="_blank">MIT License</a>
					</li>
					<li>
						CustomMill includes <a href="https://getbootstrap.com" target="_blank">Bootstrap</a> version 5.1.3. This is available under the <a href="https://github.com/twbs/bootstrap/blob/master/LICENSE" target="_blank">MIT License</a>.
					</li>
					<li>
						Syntax highlighting is implemented via <a href="https://highlightjs.org" target="_blank">highlight.js</a> licensed under the BSD 3-Clause License and the included github styling made by Vasily Polovnyov.
					</li>
					<li>
						Icons on this site are from <a href="https://fontawesome.com/" target="_blank">Font Awesome</a> which is licensed under the <a href="https://fontawesome.com/license/free" target="_blank">Font Awesome Free License</a>.
					</li></ul>
				<p>Built at 2023-02-10 16:41:13 UTC</p>
			</details>
		</footer>

		
	</body>
</html>